File: codes/naeun/34_findfirstandlastpositionofelementinsortedarray.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int[] result = new int[2];
        result[0] = binarySearchLeft(nums, target);
        result[1] = binarySearchRight(nums, target);
        return result;
    }

    private int binarySearchLeft(int[] nums, int target){
        int left = 0;
        int right = nums.length - 1;
        int result = -1;

        while(left <= right){
            int mid = left + (right - left) / 2;
            if(nums[mid] == target){
                result = mid;
                right = mid - 1; // 더 왼쪽에 있는 target을 찾기 위해
            } else if(nums[mid] < target){
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return result;
    }

    private int binarySearchRight(int[] nums, int target){
        int left = 0;
        int right = nums.length - 1;
        int result = -1;

        while(left <= right){
            int mid = left + (right - left) / 2;
            if(nums[mid] == target){
                result = mid;
                left = mid + 1; // 더 오른쪽에 있는 target을 찾기 위해
            } else if(nums[mid] < target){
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return result;
    }
}

File: codes/beomseok/34_findfirstandlastpositionofelementinsortedarray.py
class Solution:
    def binary_search(self, nums: List[int], target: int, findFirst: bool) -> int:

        start, end = 0, len(nums) - 1
        result = -1

        while start <= end:
            mid = (start + end) // 2
            if nums[mid] == target:
                result = mid
                if findFirst:
                    end = mid - 1
                else:
                    start = mid + 1
            elif nums[mid] < target:
                start = mid + 1
            else:
                end = mid - 1

        return result

    def searchRange(self, nums: List[int], target: int) -> List[int]:
        first_pos = self.binary_search(nums, target, True)
        last_pos  = self.binary_search(nums, target, False)
        return [first_pos, last_pos]